ppi.py

```diff
@@ -0,0 +1,263 @@
+    
+    return True
+
+
+def _define_descriptor(bv: BinaryView, task: BackgroundTask, descriptor_type, param) -> Optional[bool]:
+    """
+    define the descriptor type for param
+    :param bv: Binary View
+    :param task: BackgroundTask
+    :param descriptor_type: str, should be either EFI_PEI_NOTIFY_DESCRIPTOR or EFI_PEI_PPI_DESCRIPTOR
+    :param param: ParameterVariable, the descriptor parameter
+
+    :return: Optional[bool], if got cancelled or encountered an error, return False.
+        If it doesn't meet conditions, return None.
+    """
+    if not isinstance(param, MediumLevelILConstPtr):
+        return
+
+    var_addr = param.constant
+
+    var = bv.get_data_var_at(var_addr)
+    if var:
+        if descriptor_type in str(var.type):
+            # already defined
+            return
+    sym = bv.get_symbol_at(var_addr)
+
+    if sym is not None:
+        var_name = sym.name
+    else:
+        var_name = None
+
+    bv.define_user_data_var(var_addr, descriptor_type, var_name)
+    bv.update_analysis_and_wait()
+
+    var = bv.get_data_var_at(var_addr)
+    notify_descriptor = var.value
+
+    if not isinstance(notify_descriptor, dict):
+        return
+    if "Guid" not in notify_descriptor:
+        return
+
+    # define types for guid and notify entrypoint
+    protocol_name = lookup_and_define_guid(bv, notify_descriptor["Guid"])
+    if protocol_name is False:
+        return False
+
+    if "Notify" in notify_descriptor:
+        notify_entrypoint = notify_descriptor['Notify']
+        func = bv.get_function_at(notify_entrypoint)
+        if not func:
+            return
+        if not protocol_name:
+            func_name = non_conflicting_symbol_name(bv, "UnknownNotify")
+        else:
+            func_name = non_conflicting_symbol_name(bv, f"Notify{get_var_name_from_type(protocol_name)}")
+        func.type = f"EFI_STATUS {func_name}(EFI_PEI_SERVICES **PeiServices, EFI_PEI_NOTIFY_DESCRIPTOR* NotifyDescriptor, VOID* Ppi)"
+        bv.update_analysis_and_wait()
+        if not propagate_function_param_types(bv, task, func):
+            return False
+
+    return True
+
+
+def define_pei_descriptor(bv: BinaryView, task: BackgroundTask) -> bool:
+    """
+    Defines PEI related descriptors, currently this function will only define EFI_PEI_NOTIFY_DESCRIPTOR
+    and EFI_PEI_PPI_DESCRIPTOR, which may be used by PEI_SERVICES. Protocol related descriptors are not
+    supported yet.
+
+    :param bv: Binary View
+    :param task: BackgroundTask
+
+    :return: bool
+    """
+    descriptor_types = ["EFI_PEI_NOTIFY_DESCRIPTOR", "EFI_PEI_PPI_DESCRIPTOR"]
+    for descriptor_type in descriptor_types:
+        refs = list(bv.get_code_refs_for_type(descriptor_type))
+        for ref in refs:
+            if task.cancelled:
+                return False
+            instr = ref.mlil
+            if isinstance(instr, MediumLevelILCall) or isinstance(instr, MediumLevelILTailcall):
```

Comment:
There's a lot of nested logic in this function that can be avoided by inverting the conditional and using `continue`. Something like:

```python
            if not isinstance(instr, MediumLevelILCall) and not isinstance(instr, MediumLevelILTailcall):
                continue

            if not isinstance(instr.dest, MediumLevelILLoadStruct):
                continue

            # likely a call to services

            if not isinstance(instr.dest.expr_type, PointerType):
                continue

            function_type = instr.dest.expr_type.target
            params = function_type.parameters
            # and so on...
```